File: 2-medium/00417-pacific-atlantic-water-flow/solution_dfs.cpp
/*
    using two is_visited, one for atlantic ocean, one for pacific ocean

    dfs every cells that are reachable from """atlantic ocean"""
    dfs every cells that are reachable from """pacific ocean"""

    at final every cells in grid with is_visited = true in both is_visited are reachable from both side, and we can add all of them to """result"""
*/
class Solution {
    vector<vector<int>> directions{{0, -1}, {0, 1}, {1, 0}, {-1, 0}};

    int rows;
    int cols;

    void dfs(vector<vector<int>>& heights, vector<vector<bool>> &is_visited, int i, int j) {

        is_visited[i][j] = true;

        for (auto &dir: directions) {
            int new_i = i + dir[0];
            int new_j = j + dir[1];

            if (new_i >= rows || new_i < 0 || new_j >= cols || new_j < 0) {
                continue;
            }

            if (is_visited[new_i][new_j]) {
                continue;
            }

            if (heights[new_i][new_j] < heights[i][j]) {
                continue;
            }

            dfs(heights, is_visited, new_i, new_j);
        }
    }
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        rows = heights.size();
        cols = heights[0].size();

        vector<vector<bool>> atlantic_is_visited = vector(rows, vector<bool>(cols, false));
        vector<vector<bool>> pacific_is_visited = vector(rows, vector<bool>(cols, false));


        for (int i{0}; i < rows; ++i) {
            for (int j{0}; j < cols; ++j) {

                if (i == 0 || j == 0) {
                    dfs(heights, pacific_is_visited, i, j);
                }

                if (i == rows - 1 || j == cols - 1) {
                    dfs(heights, atlantic_is_visited, i, j);
                }
            }
        }

        vector<vector<int>> res{};

        for (int i{0}; i < rows; ++i) {
            for (int j{0}; j < cols; ++j) {
                if (atlantic_is_visited[i][j] && pacific_is_visited[i][j]) {
                    res.push_back(vector<int>{i, j});
                }
            }
        }

        return res;
    }
};

File: 2-medium/00417-pacific-atlantic-water-flow/solution_dfs.rs
/*
    using two is_visited, one for atlantic ocean, one for pacific ocean

    dfs every cells that are reachable from """atlantic ocean"""
    dfs every cells that are reachable from """pacific ocean"""

    at final every cells in grid with is_visited = true in "both is_visited" are reachable from both side,
    and can add all of them to """result"""
*/

impl Solution {
    fn dfs(heights: &Vec<Vec<i32>>, is_visited: &mut Vec<Vec<bool>>, directions: &[(i32, i32)], i: usize, j:usize) {

        is_visited[i][j] = true;

        let rows_i32 = heights.len() as i32;
        let cols_i32 = heights[0].len() as i32;


        for &(x, y) in directions.iter() {
            let new_i_i32 = i as i32 + x;
            let new_j_i32 = j as i32 + y;

            if new_i_i32 >= rows_i32 || new_i_i32 < 0 || new_j_i32 >= cols_i32 || new_j_i32 < 0 {
                continue;
            }

            let new_i = new_i_i32 as usize;
            let new_j = new_j_i32 as usize;

            if is_visited[new_i][new_j] {
                continue;
            }

            if heights[new_i][new_j] < heights[i][j] {
                continue;
            }

            Self::dfs(heights, is_visited, directions, new_i, new_j)
        }
    }

    pub fn pacific_atlantic(heights: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let rows = heights.len();
        let cols = heights[0].len();

        let mut pacific_is_visited = vec![vec![false; cols]; rows];
        let mut atlantic_is_visited = vec![vec![false; cols]; rows];

        let directions: [(i32, i32); 4] = [(0, -1), (0, 1), (1, 0), (-1, 0)];


        for i in 0..rows {
            for j in 0..cols {
                if i == 0 || j == 0 {
                    Self::dfs(&heights, &mut pacific_is_visited, &directions, i, j);
                }

                if i == rows - 1 || j == cols - 1 {
                    Self::dfs(&heights, &mut atlantic_is_visited, &directions, i, j);
                }
            }
        }


        let mut res: Vec<Vec<i32>> = Vec::new();

        for i in 0..rows {
            for j in 0..cols {
                if atlantic_is_visited[i][j] && pacific_is_visited[i][j] {
                    res.push(vec![i as i32, j as i32]);
                }
            }
        }

        res
    }
}
